### PR TITLE
Proof of concept

### PR DESCRIPTION
**Pull request by RobinKnipe**
Should be able to send stuff to ServiceNow

### CODE REVIEW

File: test/common.js [hunk 1]
```diff
@@ -0,0 +1,11 @@
+'use strict';
+
+global.Promise    = require('bluebird');
```
Comment (daniel-ac-martin) [inline test/common.js]: Why do you need bluebird?

Comment (daniel-ac-martin) [reply test/common.js]: i.e. Can't you just use ES6 promises?

File: index.js [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+
+// ServiceNow request
+module.exports = request
+  .post(config.endpoint)
+  .auth(config.username, config.password)
+  .type('application/json')
+  .set('Accept', 'application/json')
+  .send(config.message)
+  .then(res => {
+    console.log('Notification successfully sent, response code:', res.statusCode);
```
Comment (daniel-ac-martin) [inline index.js]: We can probably be a bit more defensive here. i.e. We always expect a 201.

Comment (daniel-ac-martin) [reply index.js]: And if we are opening a change we require an internal_id back. (Is that distinction for a later PR?)

File: .drone.yml [hunk 3]
```diff
@@ -41,4 +37,50 @@ pipeline:
...
+
+  e2e_test:
+    image: '${DOCKER_REPO}/${DOCKER_PATH}/${NAME}'
+    secrets: [ service_now_username, service_now_password ]
+    description: 'End-to-end test for the ServiceNow drone plugin. Build #${DRONE_BUILD_NUMBER}'
+    when:
+      event: push
+      branch: master
+
+  notify_slack_success:
```
Comment (daniel-ac-martin) [inline .drone.yml]: Not sure whether this should go to slack. (Or at least not the same slack channels as our current ones.)

Currently we only alert to slack on deployments.

File: .drone.yml [hunk 4]
```diff
@@ -41,4 +37,50 @@ pipeline:
...
     when:
       event: push
-      branch: master
\ No newline at end of file
+      branch: master
+
+  e2e_test:
+    image: '${DOCKER_REPO}/${DOCKER_PATH}/${NAME}'
+    secrets: [ service_now_username, service_now_password ]
+    description: 'End-to-end test for the ServiceNow drone plugin. Build #${DRONE_BUILD_NUMBER}'
```
Comment (daniel-ac-martin) [inline .drone.yml]: I don't understand what this will do. We might need a special account from the ServiceNow team for this.

**Comment by daniel-ac-martin** [comment]
Mind you. Does this need a Dockerfile update?

Also I'd like this to be runnable without docker. i.e. The docker image should contain a script that can just be called from the commandline for people that don't wish to use it as a Drone.io plugin. (That can be added later of course.)

File: test/common.js (hunk 1 above)
Comment (RobinKnipe) [reply test/common.js]: I was using `bluebird` for testing, but no longer, so it's gone...

File: index.js (hunk 2 above)
Comment (RobinKnipe) [reply index.js]: Yes, it was/is, but has now been sorted

File: .drone.yml (hunk 4 above)
Comment (RobinKnipe) [reply .drone.yml]: I don't think there's any kind of order in the SN world - I was happily chucking stuff at it the other day to try and work out what failed requests look like, and how to alert plugin users of a problem - nothing came back any other way than `201`, and I was just putting dummy values in for everything.

**Comment by RobinKnipe** [comment]
I've updated the Dockerfile - well pinched the one from LEV - the "plugin" can run inside docker, on drone, or completely autonomously as an NPM module... it'll pick up env-vars and chuck'm at SN

**Comment by RobinKnipe** [comment]
...in fact, the app can now run without `npm` - I wonder if I should change the Dockerfile... thoughts?

File: .drone.yml (hunk 4 above)
Comment (daniel-ac-martin) [reply .drone.yml]: Yeah. The only feedback I'm aware of is failed auth and missing `internal-indentifier`.

File: test/index.js [hunk 5]
```diff
@@ -0,0 +1,306 @@
...
+        password:  'snow-pass',
+        message:   '{' +
+          '"messageid":"HO_SIAM_IN_REST_CHG_POST_JSON",' +
+          '"payload":"{' +
+            '\\\"title\\\":\\\"new deployment\\\",' +
+            '\\\"endTime\\\":\\\"4000-01-01 13:09:08\\\",' +
+            '\\\"description\\\":\\\"something new\\\",' +
+            '\\\"supplierRef\\\":\\\"ext ID\\\",' +
+            '\\\"testing\\\":\\\"the test results\\\"' +
+          '}"}'
```
Comment (daniel-ac-martin) [inline test/index.js]: JSON.stringify()?

**Comment by daniel-ac-martin** [comment]
@RobinKnipe: From using `npm start`? Yes that would probably be nicer though it isn't a big deal.

It would also be nicer if the program could be configured via command line arguments (yargs?) rather than just environment variables but that could come later. (I'll raise an issue.)

**Comment by daniel-ac-martin** [comment]
#7

**Comment by daniel-ac-martin** [comment]
As this npm package provides an executable you probably want something like:
```json
  "bin": {
    "snow-change": "index.js"
},
```
in the package.json.

Which brings up naming issues. (The hardest of all programming problems.) I guess the executable shouldn't make reference to drone as it is CI agnostic.

File: .drone.yml (hunk 3 above)
Comment (RobinKnipe) [reply .drone.yml]: Yes I wasn't sure about this.

File: .drone.yml (hunk 4 above)
Comment (RobinKnipe) [reply .drone.yml]: It's ok, I've done quite a bit to handle the various possible scenarios.

**Comment by RobinKnipe** [comment]
@daniel-ac-martin - I've made the necessary changes, I think it's probably ready to be merged. Then I'll publish it and take a look at the issues...

File: config.js [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+const testingFile       = !statusUpdate && loadFromFile(pa.PLUGIN_TESTING_FILE || pa.SNOW_TESTING_FILE);
+const testing           = pa.PLUGIN_TESTING || pa.SNOW_TESTING || testingFile;
+const commentsFile      = !newDeployment && loadFromFile(pa.PLUGIN_COMMENTS_FILE || pa.SNOW_COMMENTS_FILE);
+const comments          = pa.PLUGIN_COMMENTS || pa.SNOW_COMMENTS || commentsFile;
+const deploymentOutcome = /^success$/i.test(pa.PLUGIN_DEPLOYMENT_OUTCOME || pa.status);
+const newChange         = newDeployment || !(statusUpdate || comments);
+const messageTemplates  = {
+  openChange: {
+    messageid: 'HO_SIAM_IN_REST_CHG_POST_JSON',
+    payload:   JSON.stringify(testing ? {
```
Comment (daniel-ac-martin) [inline config.js]: Do you need the ternary? If testing is undefined JSON.stringify should remove it for you anyway:
```
$ node
> obj = {
... foo: 'bar',
... baz: undefined
... };
{ foo: 'bar', baz: undefined }
> JSON.stringify(obj)
'{"foo":"bar"}'
> ⏎     
```

**Comment by daniel-ac-martin** [comment]
Found a minor way to clean up. But it can be done later if you like. Merge at will! ;-)

**Pull request merged**